File: yumi/sdk/java/src/main/java/io/yumi/SchemaBuilder.java
package io.yumi;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

/**
 * Builds the JSON tool schema sent during WebSocket registration.
 */
public final class SchemaBuilder {
    private SchemaBuilder() {}

    public static JsonObject build(RegisterOptions opts) {
        JsonObject properties = new JsonObject();
        JsonArray required = new JsonArray();

        for (ToolParameter p : opts.getParameters()) {
            JsonObject prop = new JsonObject();
            prop.addProperty("type", p.getType());
            prop.addProperty("description", p.getDescription());
            properties.add(p.getName(), prop);
            if (p.isRequired()) {
                required.add(p.getName());
            }
        }

        JsonObject params = new JsonObject();
        params.addProperty("type", "object");
        params.add("properties", properties);
        params.add("required", required);

        JsonObject function = new JsonObject();
        function.addProperty("name", opts.getName());
        function.addProperty("description", opts.getDescription());
        function.add("parameters", params);

        JsonObject schema = new JsonObject();
        schema.addProperty("type", "function");
        schema.add("function", function);

        if (opts.getTimeout() != null) {
            schema.addProperty("timeout", opts.getTimeout());
        }
        if (opts.isRequireConfirmation()) {
            schema.addProperty("require_confirmation", true);
        }
        if (opts.isAlwaysInclude()) {
            schema.addProperty("always_include", true);
        }
        if (opts.isAllowProactive()) {
            schema.addProperty("allow_proactive", true);
        }
        if (opts.isProactiveContext()) {
            schema.addProperty("proactive_context", true);
        }
        if (opts.getProactiveContextArgs() != null) {
            schema.add("proactive_context_args", new Gson().toJsonTree(opts.getProactiveContextArgs()));
        }
        if (opts.getProactiveContextDescription() != null && !opts.getProactiveContextDescription().isBlank()) {
            schema.addProperty("proactive_context_description", opts.getProactiveContextDescription());
        }

        return schema;
    }
}

File: kumi/sdk/csharp/SchemaBuilder.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Kumi;

/// <summary>
/// Builds the JSON tool schema sent during WebSocket registration.
/// </summary>
internal static class SchemaBuilder
{
    public static JsonObject Build(RegisterOptions opts)
    {
        var properties = new JsonObject();
        var required = new JsonArray();

        foreach (var p in opts.Parameters)
        {
            var prop = new JsonObject
            {
                ["type"] = p.Type,
                ["description"] = p.Description,
            };
            properties[p.Name] = prop;
            if (p.Required)
                required.Add(p.Name);
        }

        var parameters = new JsonObject
        {
            ["type"] = "object",
            ["properties"] = properties,
            ["required"] = required,
        };

        var function_ = new JsonObject
        {
            ["name"] = opts.Name,
            ["description"] = opts.Description,
            ["parameters"] = parameters,
        };

        var schema = new JsonObject
        {
            ["type"] = "function",
            ["function"] = function_,
        };

        if (opts.Timeout.HasValue)
            schema["timeout"] = opts.Timeout.Value;

        if (opts.RequireConfirmation)
            schema["require_confirmation"] = true;

        if (opts.AlwaysInclude)
            schema["always_include"] = true;

        if (opts.AllowProactive)
            schema["allow_proactive"] = true;

        if (opts.ProactiveContext)
            schema["proactive_context"] = true;

        if (opts.ProactiveContextArgs != null)
            schema["proactive_context_args"] = JsonSerializer.SerializeToNode(opts.ProactiveContextArgs);

        if (!string.IsNullOrWhiteSpace(opts.ProactiveContextDescription))
            schema["proactive_context_description"] = opts.ProactiveContextDescription;

        return schema;
    }
}
